File: ch5/python/code5_15.py
# value[i]는 길이 i인 철근의 가격입니다.
def max_value(value, N):
    # 종료 조건 : 남은 철근이 없을 때
    if N <= 0:
        return 0

    # 최대 이익. 파이썬에서 충분히 작은 값은 -float('inf')를 사용하면 됩니다.
    price = -float('inf')

    for i in range(1, N + 1):
        price = max(price, \
                      value[i] + max_value(value, N - i))
    
    return price

value = [0, 1, 5, 8, 9, 10, 17, 17, 20]
N = 8
print('길이 %d인 철근을 잘라서 팔 때 최대 %d의 이익을 얻을 수 있습니다.' % \
       (N, max_value(value, N)))
  

File: ch5/c/code5_15.c
#include <stdio.h>
#include <limits.h>

// 두 수 중 큰 수를 반환하는 도우미 함수
int getMax(int a, int b)
{
  return a > b? a: b;
}

// value[i]는 길이 i인 철근의 가격입니다.
int maxValue(int *value, int N)
{
  // 종료 조건 : 남은 철근이 없을 때 
  if(N <= 0)
    return 0;

  // 최대 이익
  int price = INT_MIN;

  for(int i = 1; i <= N; i++)
  {
    price = getMax(price, 
                   value[i] + maxValue(value, N - i));
  }
  return price;
}

int main()
{
  int value[] = {0, 1, 5, 8, 9, 10, 17, 17, 20};
  int N = 8;
  printf("길이 %d인 철근을 잘라서 팔 때 최대 %d의 이익을 얻을 수 있습니다.\n", N, maxValue(value, N));
  return 0;
}